### PR TITLE
add select and ignore flags for manifest diff command.

### PR DESCRIPTION
**Pull request by morvencao**
Resolve: https://github.com/istio/istio/issues/16178

### CODE REVIEW

File: cmd/mesh/manifest-diff.go [hunk 1]
```diff
@@ -33,11 +33,23 @@ const YAMLSuffix = ".yaml"
 type manifestDiffArgs struct {
 	// compareDir indicates comparison between directory.
 	compareDir bool
+	// selection specifies the resources to compare, it's comma-separated of resource indicators
+	// in the form of "<kind>:<namespace>:<name>,<kind>:<namespace>:<name>".
```
Comment (ostromart) [inline cmd/mesh/manifest-diff.go]: suggest changing names to selectResources and ignoreResources, as selection is a bit confusing.
also, i propose the comment to read "selectResources constrains the list of resources to compare to only the ones in this list, ignoring all others. The format of each list item is <kind>:<namespace>:<name> and the items are comma separated. The * character represents wildcard selection. e.g. 
Deployment:istio-system:* - compare all deployments in istio-system namespace
Service:*:istio-pilot - compare Services called "istio-pilot" in all namespaces.

File: cmd/mesh/manifest-diff.go [hunk 2]
```diff
@@ -33,11 +33,23 @@ const YAMLSuffix = ".yaml"
...
+	selection string
+	// ignore specifies the resources to ignore, it's comma-separated of resource indicators
+	// in the form of "<kind>:<namespace>:<name>,<kind>:<namespace>:<name>".
```
Comment (ostromart) [inline cmd/mesh/manifest-diff.go]: ignoreResources ignores all listed items during comparison. It uses the same list format as selectResources.

File: cmd/mesh/manifest-diff.go [hunk 3]
```diff
@@ -33,11 +33,23 @@ const YAMLSuffix = ".yaml"
...
+	ignore string
 }
 
 func addManifestDiffFlags(cmd *cobra.Command, diffArgs *manifestDiffArgs) {
 	cmd.PersistentFlags().BoolVarP(&diffArgs.compareDir, "directory", "r",
 		false, "compare directory")
+	cmd.PersistentFlags().StringVar(&diffArgs.selection, "select", "::",
+		"specifies the resources to compare, it's comma-separated of resource indicators"+
+			"in the form of \"<kind>:<namespace>:<name>,<kind>:<namespace>:<name>\".")
```
Comment (ostromart) [inline cmd/mesh/manifest-diff.go]: Use the text above, if it looks good to you.

File: cmd/mesh/manifest-diff.go [hunk 4]
```diff
@@ -33,11 +33,23 @@ const YAMLSuffix = ".yaml"
...
+	cmd.PersistentFlags().StringVar(&diffArgs.ignore, "ignore", "",
+		"specifies the resources to ignore, it's comma-separated of resource indicators"+
+			"in the form of \"<kind>:<namespace>:<name>,<kind>:<namespace>:<name>\".")
```
Comment (ostromart) [inline cmd/mesh/manifest-diff.go]: same here.

File: cmd/mesh/manifest-diff.go [hunk 5]
```diff
@@ -47,17 +59,20 @@ func manifestDiffCmd(rootArgs *rootArgs, diffArgs *manifestDiffArgs) *cobra.Comm
 		Long:  "The diff-manifest subcommand is used to compare manifest from two files or directories.",
 		Args:  cobra.ExactArgs(2),
 		Run: func(cmd *cobra.Command, args []string) {
+			if diffArgs.selection != "::" && diffArgs.ignore != "" {
+				logAndFatalf(rootArgs, "Cannot specify both the selected and ignored resources.")
+			}
```
Comment (ostromart) [inline cmd/mesh/manifest-diff.go]: I don't think these need to be mutually exclusive as long as the order they are applied is documented. I'd propose select before ignore. 
In that case you could do --selectResources Deployment:*:* --ignoreResources *:*:istio-pilot
That would compare all Deployments except the one called istio-pilot.

File: pkg/object/objects.go [hunk 6]
```diff
@@ -410,6 +411,123 @@ func ManifestDiff(a, b string) (string, error) {
...
 
+// buildResourceRegexp translates the resource indicator to regexp.
+func buildResourceRegexp(s string) *regexp.Regexp {
+	hash := strings.Split(s, ":")
+	for i, v := range hash {
+		if v == "" {
+			hash[i] = ".*"
+		}
+	}
+	return regexp.MustCompile(strings.Join(hash, ":"))
```
Comment (ostromart) [inline pkg/object/objects.go]: This will panic if it fails so we can't use it here.

File: pkg/object/objects.go [hunk 7]
```diff
@@ -410,6 +411,123 @@ func ManifestDiff(a, b string) (string, error) {
...
+				}
+			}
+		}
+	}
+	return sb.String(), err
+}
+
+// ManifestDiffWithIgnore checks the manifest differences with ignored indicator.
+func ManifestDiffWithIgnore(a, b, ignore string) (string, error) {
+	ao, err := ParseK8sObjectsFromYAMLManifest(a)
```
Comment (ostromart) [inline pkg/object/objects.go]: You could restructure the code as follows:
1. create a top level function called ManifestDiffWithSelectAndIgnore(...)
2. get objects slices for manifests a and b.
3. apply select and ignore filters to those lists
4. create the aom, bom maps and proceed as before.

File: cmd/mesh/manifest-diff.go (hunk 1 above)
Comment (morvencao) [reply cmd/mesh/manifest-diff.go]: done.

File: cmd/mesh/manifest-diff.go (hunk 2 above)
Comment (morvencao) [reply cmd/mesh/manifest-diff.go]: done.

File: cmd/mesh/manifest-diff.go (hunk 3 above)
Comment (morvencao) [reply cmd/mesh/manifest-diff.go]: done.

File: cmd/mesh/manifest-diff.go (hunk 4 above)
Comment (morvencao) [reply cmd/mesh/manifest-diff.go]: done.

File: cmd/mesh/manifest-diff.go (hunk 5 above)
Comment (morvencao) [reply cmd/mesh/manifest-diff.go]: done.

File: pkg/object/objects.go (hunk 6 above)
Comment (morvencao) [reply pkg/object/objects.go]: Yes, changed to `regexp.Compile(strings.Join(hash, ":"))`

File: pkg/object/objects.go (hunk 7 above)
Comment (morvencao) [reply pkg/object/objects.go]: done.

File: pkg/object/objects_test.go [hunk 8]
```diff
@@ -732,3 +732,417 @@ func TestUnmarshalWithJSONPB(t *testing.T) {
...
+      - name: citadel
+        image: docker.io/istio/citadel:1.1.8
+        args:
+        - "--append-dns-names=true"
+        - "--grpc-port=8060"
+        - "--grpc-hostname=citadel"
+        - "--citadel-storage-namespace=istio-system"
+        - "--custom-dns-names=istio-pilot-service-account.istio-system:istio-pilot.istio-system"
+        - "--monitoring-port=15014"
+        - "--self-signed-ca=true"
```
Comment (ostromart) [inline pkg/object/objects_test.go]: having this much stuff in the resource actually makes it harder to know what the test is doing and what is expected. i recommend defining very minimal resources so it's clear what the differences are.

File: pkg/object/objects_test.go [hunk 9]
```diff
@@ -732,3 +732,417 @@ func TestUnmarshalWithJSONPB(t *testing.T) {
...
+      - name: citadel
+        image: docker.io/istio/citadel:1.2.2
+        args:
+        - "--append-dns-names=true"
+        - "--grpc-port=8060"
+        - "--grpc-hostname=citadel"
+        - "--citadel-storage-namespace=istio-system"
+        - "--custom-dns-names=istio-pilot-service-account.istio-system:istio-pilot.istio-system"
+        - "--monitoring-port=15014"
+        - "--self-signed-ca=true"
```
Comment (ostromart) [inline pkg/object/objects_test.go]: same here, i recommend paring these resources back to be as small as possible to still be able to get a diff for the test.

**Comment by morvencao** [comment]
@ostromart OK, I'll fix the UT in the followup PR.

**Pull request merged**